x/babylon/keeper/handler_plugin.go

```diff
@@ -64,11 +67,35 @@ func (h CustomMsgHandler) DispatchMsg(ctx sdk.Context, contractAddr sdk.AccAddre
 		return nil, nil, nil, sdkerrors.ErrUnauthorized.Wrapf("contract has no permission for Babylon operations")
 	}
 
-	return h.handleTestMsg(ctx, contractAddr, customMsg.Test)
+	return h.handleMintRewardsMsg(ctx, contractAddr, customMsg.MintRewards)
 }
 
-func (h CustomMsgHandler) handleTestMsg(ctx sdk.Context, actor sdk.AccAddress, testMsg *contract.TestMsg) ([]sdk.Event, [][]byte, [][]*codectypes.Any, error) {
-	return []sdk.Event{}, nil, nil, nil
+func (h CustomMsgHandler) handleMintRewardsMsg(ctx sdk.Context, actor sdk.AccAddress, _ *contract.MintRewardsMsg) ([]sdk.Event, [][]byte, [][]*codectypes.Any, error) {
```

Comment:
Yeah basically this can be invoked inside `BeginBlock`

Reply:
I'm not so sure. What if the FP set is inactive? What if they are not submitting signatures? Etc.

Reply:
What is the problem with this design? Do you think it's too complicated?

In my opinion, it's better to mint rewards related to block finalisation only when such blocks are being effectively finalised. But I can refactor, if we agree it's worth it.

Reply:
Those contract <-> blockchain messages are very efficient by the way.

Reply:
Distribution is an entirely different matter.

For distribution, we need to take into account the active FP set. But there's no need to do that at every block, it can be epoch based. With a small epoch, that will introduce a small error, but improve efficiency and reduce the load. It will also help reduce rounding errors.

Reply:
This is in fact being executed at the end blocker. The current sequence is `blockchain -> end_block -> contract -> blockchain`.

Reply:
As per offline discussion, it is necessary to mint tokens only upon finalising a block. If the protocol mints tokens only upon BeginBlock, then it's possible that the finality contract admin moves all the inflation at wish, e.g., rugpull, bridging it to other contracts, etc.. This is a security critical issue. Babylon does not have this problem because it 1) the inflation is in the fee collector account controlled by gov prop, rather than sth like a contract admin, and 2) people only recognise BTC stake finalised prefix as finalised, but this might not be the case in consumer chains.

Reply:
After discussion, closed in favour of #62.

Reply:
Yes. ADR is to come.